syntax = "proto3";

package jrtransport;

option cc_generic_services = false;
option cc_enable_arenas = false;

message Error {
    bool exists = 1;
    string message = 2;
}

message Contact {
    string hash = 1;
    string address = 2;
    string port = 3;
}

message KeyValue {
    string key = 1;
    bytes value = 2;
}

message PingRequest {
    Contact sender = 1;
}

message PongResponse {
    Contact sender = 1;
    Error error = 2;
}

message AuthRequest {
    Contact source = 1;
    string secret = 2;
}

message AuthResponse {
    Contact destination = 1;
    Error error = 2;
}

message StoreRequest {
    Contact sender = 1;
    KeyValue element = 2;
}

message StoreResponse {
    Contact sender = 1;
    Error error = 2;
}

message FindNodeRequest {
    Contact sender = 1;
    string hash = 2;
}

message FindNodeResponse {
    Contact sender = 1;
    Error error = 2;
    repeated Contact nodes = 3;
}

message FindValueRequest {
    Contact sender = 1;
    string hash = 2;
}

message FindValueResponse {
    Contact sender = 1;
    Error error = 2;

    enum ResponseType {
        VALUE_FOUND = 0;
        VALUE_NOT_FOUND = 1;
    }
    ResponseType type = 3;
    KeyValue element = 4;
    repeated Contact nodes = 5;
}

service JRTransportService {
    rpc Ping (PingRequest) returns (PongResponse) {}
    rpc Auth (AuthRequest) returns (AuthResponse) {}

    rpc Store (StoreRequest) returns (StoreResponse) {}

    rpc FindNode (FindNodeRequest) returns (FindNodeResponse) {}
    rpc FindValue (FindValueRequest) returns (FindValueRequest) {}
}